Products/ZenUtils/Executor.py

```diff
@@ -1,84 +1,396 @@
+            )
+
+    def error(self, failure):
+        if self._ontimeout:
+            self._ontimeout.cancel()
+        if failure.check(_ShutdownException):
+            return
+        if not self.deferred.called:
+            self.deferred.errback(failure)
+        else:
+            self._log.debug(
+                "Failure ignored  task-id=%s reason=timeout", self.id,
+            )
+
+    def _timeout(self, failure, timeout):
+        self._log.debug("Task timed out  task-id=%s", self.id)
+        self.deferred.errback(defer.TimeoutError("timeout"))
+        self._ontimeout = None
+
+    def __repr__(self):
+        return (
+            "<{0.__module__}.{0.__name__} id={1}>"
+        ).format(type(self), self.id)
+
+
+class _ShutdownException(Exception):
+    pass
+
+
+def _shutdown_executor():
+    raise _ShutdownException()
+
+
+class DynamicDeferredSemaphore(object):
+    """A semaphore that supports runtime adjustments to its size.
+
+    Setting the DynamicDeferredSemaphore to a zero limit effectively
+    creates a semaphore with no limit.
+
+    @ivar limit: The total number of tokens available for acquisition.
+    @type limit: int
+
+    @ivar tokens: The number of tokens available for acquisition.
+    @type tokens: int
     """
-    def __init__(self, deferred, callable, *args, **kw):
-        self._callable = callable
-        self._args = args
-        self._kw = kw
-        self._deferred = deferred
-        
-    def __call__(self):
-        deferred =  maybeDeferred(self._callable,*self._args, **self._kw)
-        deferred.addCallback(self._finished)
-        deferred.addErrback(self._error)
-        
-    def _finished(self, result):
-        self._deferred.callback(result)
-        
-    def _error(self, result):
-        self._deferred.errback(result)
+
+    def __init__(self, limit):
+        """Initialize a DynamicDeferredSemaphore instance.
+
+        @param limit: total number of availale tokens.
+        @type tokens: int
+        """
+        if limit < 0:
+            raise ValueError("Cannot set the limit to a negative value")
+        self._tokens = limit
+        self._limit = limit
+        self._waiting = []
+
+    @property
+    def tokens(self):
+        return max(self._tokens, 0)
+
+    @property
+    def limit(self):
+        return self._limit
+
+    @limit.setter
+    def limit(self, limit):
+        """Update the total number of tokens available for acquisition.
+
+        Reducing the limit does not cancel acquired tokens.
+
+        Increasing the limit will immediately fire pending acquisitions.
+        Reducing the limit to zero will also fire pending acquisitions.
+        """
+        if limit < 0:
+            raise ValueError("Cannot set the limit to a negative value")
+        diff = limit - self._limit
+        self._limit = limit
+        self._tokens = self._tokens + diff
+        if self._limit == 0:
+            while self._waiting:
```

Comment:
self._tokens should probably also get set to 0 here, otherwise you might get in this state:
limit=10
acquire 1 token
limit = 0 (_tokens is now -9)
limit = 3 (_tokens is now -6)
release the only token (_tokens is now -5)
Cannot acquire any more tokens
(This might make a good test case)

Reply:
I'll add a test and fix up the code to resolve any failure.

Reply:
The test came out a little different:
limit=10
acquire 1 token
limit=0 (_tokens is -1; i.e. 9 + -10)
limit=3 (_tokens is 2)
release the token (_tokens is now 3)

However, I wonder what the desired behavior should be when setting the limit to zero (infinite) in the presense of unreleased tokens.